### PR TITLE
net: lwm2m: Add LWM2M object 9 software management

### PR DESCRIPTION
**Pull request by MrOneTwo**
Adds support for LWM2M object 9 Software management.
This is implemented according to this release:
http://openmobilealliance.org/release/LWM2M_SWMGMT/V1_0_1-20200616-A/

Note that the XML is lacking some resources and for that reason those
resources are not included. This is a known problem by OMA and will be
fixed in a later releases.

This uses the lwm2m_pull_context to pull binaries in case
FIRMWARE_PULL_SUPPORT is enabled

This PR has been introduced before by @DunderRoffe but he closed it because he couldn't finalize it (lack of time). I'm trying to pick up the torch. There has been a few small bugs fixed between then and this PR.

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:610]: Is the PR complete? I don't see this being defined anywhere.

**Comment by rlubos** [comment]
It seems that we have an alternative proposal in https://github.com/zephyrproject-rtos/zephyr/pull/40628. 
Before we proceed, we need to settle which PR should take precedence, both need some work and there's no point addressing issues in both of them since only one can get in.

@MrOneTwo @jarlamsa What's your opinion on this? I'm asking specifically which one of you is more determined to put effort into addressing review comments and proceeding with the PR.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:610]: It's not. I've realized there are commits that this depends on and I'm working on integrating them into this PR. Sorry for the confusion.

**Comment by MrOneTwo** [comment]
@rlubos @jarlamsa I'll try to address the issues in this PR during the coming week. All I can say is that I'm pretty committed to getting this in soon.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:22]: Woudl be good to use `ARG_UNUSED()` on `context` here.

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_pull_context.c:366]: Isn't it a bit of a waste, to keep separate PULL context per each object instance, but then limit the usage to a single download at once (note that this structure is pretty large due to fields like `uri` or `lwm2m_ctx`)?  Perhaps it'd be more economic to have a single context, and then verify inside `lwm2m_pull_context_start_transfer()` if it's in use (it'd then be the caller's responsibility to try again later).

File: include/net/lwm2m.h
Comment (rlubos) [inline include/net/lwm2m.h:383]: I think `lwm2m_engine_execute_cb_t` would be more appropriate here. Same for other actions (deactivate/install/uninstall).

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:44]: It's a lot of space for the package name/version, needed per object instance. I think it'd be good to make those values configurable with Kconfig.

File: include/net/lwm2m.h
Comment (rlubos) [inline include/net/lwm2m.h:434]: I just wonder, is this really needed? There's no special action needed on the object implementation side before the callback is exectued, so the app could simply use generic `lwm2m_engine_register_read_callback()`.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:200]: Extra empty line. There are other places as well like this, in general we don't put more than one empty line between instructions in the code.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:734]: You should also set version_major/minor fields.

**Review by rlubos — COMMENTED** [review]
Initial high level feedback. Still need to take a closer look at the SW managmament state machine, but it looks pretty good already.

**Comment by rlubos** [comment]
@MrOneTwo Thanks again for the contribution, we've agreed with @jarlamsa to focus on this PR since it brings more functionality to Zephyr (i.e. PULL mode support).

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:22]: The function is now a bit misleading as setting update result from error when there is no error suggests something else.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:577]: There should be some possibility to have an external function doing the package integrity verification and reporting the results to the object. Currently I can't see a clear path leading here.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:570]: There isn't `firmware_pull_context` in `lwm2m_swmgmt_data` when `CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT` isn't enabled. Is this supposed only to be used with the PULL-method? What about PUSH-method?

**Review by jarlamsa — COMMENTED** [review]
Thank you for the contribution. Here is an initial review before adapting our own solution to this object implementation. Might be that something arises when I start trying it out but I'll let you know.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:577]: When you say `external` do you mean something settable by the Zephyr's user? A callback that they can set or should it be a static function in the lwm2m scope? Is it fine if we leave that for a future PR or do you think this has to be a part of this PR?

I'd obviously like it to do this but this PR is already pretty big. The comment needs fixing for sure if we choose to implement this later.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:570]: This doesn't support PUSH atm. This is something that should be implemented in the future. Any safety check that we could introduce here?

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:200]: fixed

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:361]: fixed

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:559]: That's a bit unfortunate to be honest. Do  Iunderstand correctly, that the only limitation is the PULL context provided to the result callback? Perhaps instead of providing this pointer, we could simple feed the callback with the instance number? On the FW update object side it's not used anyway, and here we only use the pointer to obtain the acual object instance - we could achieve the same with instance number and `find_index()`

Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:559]: Lack of PUSH is indeed unfortunate. Any estimates that how long it would take to implement?

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c]: Should be good now

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:366]: Agree. I'll go single context.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:559]: Since I've inherited this PR I cannot answer how long it would take, right now. I'm just not knowledgeable enough about this topic. I'll do my best to do my homework and see if I can implement it in a reasonable time.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:745]: Setting it to 1.1

Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:745]: Why 1.1? I see in [OMA registry](https://technical.openmobilealliance.org/OMNA/LwM2M/LwM2MRegistry.html) that this object is only available in version 1.0.

**Review by DunderRoffe — CHANGES_REQUESTED** [review]
Nice to see someone picking up my slack :)

I appreciate that you want to give me credit but as I am not the one doing this and or testing it I would appreciate if you would remove my name and sign-of-by line from the commits.

**Comment by MrOneTwo** [comment]
@DunderRoffe Will do. Didn't forget, just didn't get to it yet :) Thanks for the comments, I'll start fixing that stuff soon.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:25]: Hmmm @jarlamsa I understand but since it's a static function of this file do you think this is a problem? I think it's pretty readable that if no error is present then we report success. Maybe we can make the name more generic? `set_update_result`? I'm open to suggestions :)

@rlubos Added `ARG_UNUSED()`

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:362]: @rlubos I would appreciate you taking a look at this solution https://github.com/zephyrproject-rtos/zephyr/pull/40552/commits/afbb37219139acd2467602f543c850be9e6ed23e - this is handling failure when a package is already being pulled.

Safety check if the work is already scheduled might be unnecessary but I'd really like your input on this.

File: include/net/lwm2m.h
Comment (MrOneTwo) [reply include/net/lwm2m.h]: Fixed

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:362]: I just wonder, do we really need `firmware_work` in this case? This initial check to determine if download is already ongoing looks fine, but I don't see why would need a work here, `firmware_transfer()` could just be a regular function called from that context.

I'd rather see a delayed work use on the caller side - if the function returns `-EALREADY` it should set a delayed work and retry later.

I also don't think we should expect `struct firmware_pull_context *` as an argument for the `lwm2m_pull_context_start_transfer()` - since the only instance in use would exist in the `lwm2m_pull_context.c`, the callers don't even need to be aware of this structure. We only need to provide a few information needed to start the download - URI string pointer, instance ID, callback pointers, perhaps nicely wrapped into some helper struct. But this struct does not need to be in the static storage, it could exist on a stack, since all of the data would be copied into the pull context when the download is started.

Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:362]: Hard to disagree with anything you've pointed out here.

I'll get to implementing these changes.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:25]: `set_update_result` would be good name for this

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: I've implemented this but I did per download request. Do you see any value in storing a verify function per object instance?

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: done and I'm setting some lower default values

File: include/net/lwm2m.h
Comment (MrOneTwo) [reply include/net/lwm2m.h:457]: Hmmm... on a functional level I don't think this is needed - meaning this doesn't provide unique functionality compared to `lwm2m_engine_register_read_callback`. The thing is that *read package* callback reads the resource's version attribute. It's not a typical *read the data callback*. So technically the user can implement reading the version attribute and pass that via the `lwm2m_engine_register_read_callback` but I think that would be a bit misleading. Having a separate function creates a distinction between *set read data callback* and *set read package version attribute*.

I hope I'm getting this right ;)

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:23]: fixed

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: I thought about this and I think I should not implement the PUSH method. That's because I have no way to test it in the foreseeable future. I wouldn't feel comfortable with merging this without testing it. I might be the one who implements that in the future but I can't do it right now.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:588]: Could you elaborate why? For setting a correct initial state?

**Comment by MrOneTwo** [comment]
@rlubos I think this is ready for you to take a look. I hope it's not a problem that I squashed the commits (if you have any pointers on how to make the review process easier then I'm all ears).

**Comment by linusnielsen** [comment]
There is a small issue with this implementation of Object 9. There is no way to uninstall/delete the package after a reboot, because the state is always `UPD_STATE_INITIAL` at boot. There should probably be a way to set the initial state to `UPD_STATE_INSTALLED` in those cases where the package is already installed. Perhaps a callback? Or should the application use `lwm2m_engine_set_u8("/9/x/7", UPD_STATE_INSTALLED)` in these cases?

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:631]: Not sure here, this will result in an error response being sent to the server if some other download is in progress, wouldn't that be confusing for the server side (all in all URI write would succeed).

Shouldn't we rather have an extra delayed work item (per instance) and set it to retry download after some time in case of `EALREADY`?

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:73]: I think this object should implement the same logic for handling start_transfer failures as SW mgmt.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:180]: This doesn't seem to be used at all (only set to NULL in one place), do we need it? I'd rather avoid introducing dead code, it might be confusing for folks analyzing the code.

**Review by rlubos — COMMENTED** [review]
Thanks, I've left a few new comments, but overall I think we're pretty good here, nothing major. Good work with the rework, thanks!

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (linusnielsen) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:180]: It is called on line 582. However, I believe that callback should be part of `struct lwm2m_swmgmt_data`.

Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:180]: Well, I'd rather say the pointer is used, not called - it won't be called as it's only set to NULL in one place. And there's no API to change it value. But I agree with the latter, if it stays it should be part of the struct.

File: subsys/net/lib/lwm2m/Kconfig
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/Kconfig:269]: fixed

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:37]: done

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_pull_context.c:24]: done

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:180]: I'm not sure about this. Why do we want to store a verify callback per object instance? What's the benefit there?
When I was porting object 9, the idea was to hold a verify callback per pull request. Mostly because this seems to do the job and is cheap memory wise. The API for that is the `struct requesting_object` - that's where one could set the `verify_cb`. Currently it's not really a user facing API but it conveys the idea of not storing a verification callback persistently, per object instance.

Comment (linusnielsen) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:180]: My thinking is that the verification might differ depending on the object requesting the pull. Each object might have a different way of checking the integrity, depending on the contents of the package.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: done

File: include/net/lwm2m.h
Comment (MrOneTwo) [reply include/net/lwm2m.h:403]: fixed. Please double check if the documentation is correct and understandable.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:497]: I'll delete this comment. I don't know what was the original author's intention here.

Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: deleted

File: include/net/lwm2m.h
Comment (MrOneTwo) [reply include/net/lwm2m.h:403]: Changed to `lwm2m_swmgmt_set_read_package_version_cb`

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:626]: That's a good question... should we? Adding a delayed work would bump up the memory usage a bit for sure. Not sure about this one either.

File: subsys/net/lib/lwm2m/Kconfig
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:269]: nit: We typically avoid `default n` since it's the implicit value.

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:626]: I don't have a strong opinion here, I guess we coudl leave it as it is now (i. e. report an error to the server), and in cases it turns out problematic update in the future.

File: subsys/net/lib/lwm2m/Kconfig
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/Kconfig:269]: I'm a bit confused here. This change is in response to this comment https://github.com/zephyrproject-rtos/zephyr/pull/40552#discussion_r768668057
Is there a different way to turn this off without `default n`?

Comment (jarlamsa) [reply subsys/net/lib/lwm2m/Kconfig:269]: Removing that `default n` line will result as same behaviour.

Comment (rlubos) [reply subsys/net/lib/lwm2m/Kconfig:269]: Just delete the `default n` from the option, `default n` is the same as if there's no default at all, so since it's redundant we avoid it (tree-wise, it's not an lwm2m thing).

Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/Kconfig]: ok, fixed

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (linusnielsen) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:8]: Typo: Should be `CONFIG_LWM2M_LOG_LEVEL_DBG`

Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: And I do typing for a living 🤦 thanks
I changed this to `CONFIG_LWM2M_LOG_LEVEL`. It shouldn't be the `DBG` by default.

**Review by rlubos — APPROVED** [review]
No more comments from my side, thanks

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c
Comment (MrOneTwo) [reply subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c]: Since PUSH means just pushing data into the Software Management object, this should be supported.

**Review by rerickson1 — APPROVED** [review]
Looks great. Object 9 is on our list of requirements for our next project, looking forward to trying this out.

**Comment by carlescufi** [comment]
@DunderRoffe @linusnielsen @jarlamsa could you please take another look and approve if appropriate?

**Review by linusnielsen — APPROVED** [review]

**Comment by carlescufi** [comment]
Please do not merge until @DunderRoffe has had a chance to take another look.

**Review by DunderRoffe — COMMENTED** [review]
Commit messages and sign off looks good

Copyright notice from the firmware update object should probably still be added

**Comment by MrOneTwo** [comment]
@DunderRoffe I looked through the code and the `lwm2m_obj_firmware_pull` code was mainly moved into the `lwm2m_pull_context`. Those two have the same copyright header. Are you saying I should add the `lwm2m_obj_firmware_pull` copyright header in the `lwm2m_obj_swmgmt.c`? Because that code seems to mostly be unique?

**Review by DunderRoffe — APPROVED** [review]

**Pull request merged**
